### PR TITLE
Cluster Governance Pallet

### PR DESCRIPTION
**Pull request by yahortsaryk**
### Description

This PR introduces the Cluster Governance feature. Previously, only the SUDO account was authorized to create an active cluster in the Cere Network. Now, anyone can create an inactive cluster, bond tokens for that cluster, add genesis nodes, invite external nodes, and later initiate a proposal to the network to activate the cluster with certain protocol parameters that impact economics and distribution of rewards between Cere token holders.

In sake for the overview, I'll list the high-level changes the current PR contains:
- Bonding for clusters is introduced
- Cluster Node Kind type is introduced (`Genesis`, `External`)
- Cluster Node Status is introduced (`AwaitsValidation`, `ValidationSucceeded`, `ValidationFailed`)
- Cluster Local Referendum is introduced where cluster members can agree on an initial set of protocol parameters before propagating the proposal to the network
- Cluster Public Referendum (integrated with the OpenGov) is introduced where the network participants can accept or reject changes in a cluster protocol parameters proposed by its members.
- `ClusterGovParams` type was renamed to `ClusterProtocolParams` to ensure ubiquitous naming and better reflect the type purpose. Also, internal some traits such as `ClusterVisitor` were split and renamed to more domain-specific types.
- Tests and benchmarks for new and changed functions are added
- Migrations for old data structures are provided

### Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Dependency upgrade (A change in substrate or any 3rd party crate version)

### Migrations and Hooks

- [x] This change requires a runtime migration.
- [ ] Modifies `on_initialize`
- [ ] Modifies `on_finalize`

### Checklist

- [x] Change has been tested locally.
- [x] Change adds / updates tests.
- [x] Changelog doc updated.

### CODE REVIEW

File: pallets/ddc-clusters/src/weights.rs [hunk 1]
```diff
@@ -31,108 +31,168 @@ pub trait WeightInfo {
...
+	// Proof Skipped: DdcClusters ClustersNodes (max_values: None, max_size: None, mode: Measured)
+	// Storage: unknown `0x89eb0d6a8a691dae2cd15ed0369931ce0a949ecafa5c3f93f8121833646e15c3` (r:1 w:0)
+	// Proof Skipped: unknown `0x89eb0d6a8a691dae2cd15ed0369931ce0a949ecafa5c3f93f8121833646e15c3` (r:1 w:0)
+	// Storage: unknown `0xc3ad1d87683b6ac25f2e809346840d7a7ed0c05653ee606dba68aba3bdb5d957` (r:1 w:0)
+	// Proof Skipped: unknown `0xc3ad1d87683b6ac25f2e809346840d7a7ed0c05653ee606dba68aba3bdb5d957` (r:1 w:0)
 	fn add_node() -> Weight {
-		Weight::from_parts(599_000_000_u64, 0)
-			.saturating_add(T::DbWeight::get().reads(14_u64))
-			.saturating_add(T::DbWeight::get().writes(5_u64))
+		Weight::from_parts(3_972_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/weights.rs:82]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

File: pallets/ddc-clusters/src/weights.rs [hunk 2]
```diff
@@ -31,108 +31,168 @@ pub trait WeightInfo {
...
 impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 	// Storage: DdcClusters Clusters (r:1 w:1)
+	// Proof Skipped: DdcClusters Clusters (max_values: None, max_size: None, mode: Measured)
+	// Storage: DdcClusters ClustersNodesStats (r:0 w:1)
+	// Proof Skipped: DdcClusters ClustersNodesStats (max_values: None, max_size: None, mode: Measured)
 	// Storage: DdcClusters ClustersGovParams (r:0 w:1)
+	// Proof Skipped: DdcClusters ClustersGovParams (max_values: None, max_size: None, mode: Measured)
 	fn create_cluster() -> Weight {
-		Weight::from_parts(15_000_000_u64, 0)
+		Weight::from_parts(350_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/weights.rs:47]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

File: pallets/ddc-clusters/src/weights.rs [hunk 3]
```diff
@@ -31,108 +31,168 @@ pub trait WeightInfo {
...
+	// Proof Skipped: DdcNodes StorageNodes (max_values: None, max_size: None, mode: Measured)
+	// Storage: DdcClusters ClustersNodes (r:1 w:1)
+	// Proof Skipped: DdcClusters ClustersNodes (max_values: None, max_size: None, mode: Measured)
+	// Storage: DdcClusters ClustersNodesStats (r:1 w:1)
+	// Proof Skipped: DdcClusters ClustersNodesStats (max_values: None, max_size: None, mode: Measured)
 	fn remove_node() -> Weight {
-		Weight::from_parts(24_000_000_u64, 0)
-			.saturating_add(T::DbWeight::get().reads(2_u64))
-			.saturating_add(T::DbWeight::get().writes(2_u64))
+		Weight::from_parts(505_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/weights.rs:95]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

File: pallets/ddc-clusters/src/weights.rs [hunk 4]
```diff
@@ -31,108 +31,168 @@ pub trait WeightInfo {
...
+			.saturating_add(T::DbWeight::get().reads(4_u64))
+			.saturating_add(T::DbWeight::get().writes(3_u64))
 	}
 	// Storage: DdcClusters Clusters (r:1 w:1)
+	// Proof Skipped: DdcClusters Clusters (max_values: None, max_size: None, mode: Measured)
 	fn set_cluster_params() -> Weight {
-		Weight::from_parts(16_000_000_u64, 0)
+		Weight::from_parts(244_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/weights.rs:102]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

File: pallets/ddc-clusters/src/weights.rs [hunk 5]
```diff
@@ -31,108 +31,168 @@ pub trait WeightInfo {
...
-			.saturating_add(RocksDbWeight::get().writes(2_u64))
+		Weight::from_parts(505_000_000_u64, 0)
+			.saturating_add(RocksDbWeight::get().reads(4_u64))
+			.saturating_add(RocksDbWeight::get().writes(3_u64))
 	}
 	// Storage: DdcClusters Clusters (r:1 w:1)
+	// Proof Skipped: DdcClusters Clusters (max_values: None, max_size: None, mode: Measured)
 	fn set_cluster_params() -> Weight {
-		Weight::from_parts(16_000_000_u64, 0)
+		Weight::from_parts(244_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/weights.rs:183]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

File: pallets/ddc-clusters/src/weights.rs [hunk 6]
```diff
@@ -31,108 +31,168 @@ pub trait WeightInfo {
...
+	// Proof Skipped: DdcClusters ClustersNodes (max_values: None, max_size: None, mode: Measured)
+	// Storage: unknown `0x89eb0d6a8a691dae2cd15ed0369931ce0a949ecafa5c3f93f8121833646e15c3` (r:1 w:0)
+	// Proof Skipped: unknown `0x89eb0d6a8a691dae2cd15ed0369931ce0a949ecafa5c3f93f8121833646e15c3` (r:1 w:0)
+	// Storage: unknown `0xc3ad1d87683b6ac25f2e809346840d7a7ed0c05653ee606dba68aba3bdb5d957` (r:1 w:0)
+	// Proof Skipped: unknown `0xc3ad1d87683b6ac25f2e809346840d7a7ed0c05653ee606dba68aba3bdb5d957` (r:1 w:0)
 	fn add_node() -> Weight {
-		Weight::from_parts(599_000_000_u64, 0)
-			.saturating_add(RocksDbWeight::get().reads(14_u64))
-			.saturating_add(RocksDbWeight::get().writes(5_u64))
+		Weight::from_parts(3_972_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/weights.rs:163]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

File: pallets/ddc-clusters/src/weights.rs [hunk 7]
```diff
@@ -31,108 +31,168 @@ pub trait WeightInfo {
...
 impl WeightInfo for () {
 	// Storage: DdcClusters Clusters (r:1 w:1)
+	// Proof Skipped: DdcClusters Clusters (max_values: None, max_size: None, mode: Measured)
+	// Storage: DdcClusters ClustersNodesStats (r:0 w:1)
+	// Proof Skipped: DdcClusters ClustersNodesStats (max_values: None, max_size: None, mode: Measured)
 	// Storage: DdcClusters ClustersGovParams (r:0 w:1)
+	// Proof Skipped: DdcClusters ClustersGovParams (max_values: None, max_size: None, mode: Measured)
 	fn create_cluster() -> Weight {
-		Weight::from_parts(15_000_000_u64, 0)
+		Weight::from_parts(350_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/weights.rs:128]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

File: pallets/ddc-clusters/src/weights.rs [hunk 8]
```diff
@@ -31,108 +31,168 @@ pub trait WeightInfo {
...
-		Weight::from_parts(17_000_000_u64, 0)
-			.saturating_add(T::DbWeight::get().reads(1_u64))
-			.saturating_add(T::DbWeight::get().writes(1_u64))
+	// Proof Skipped: DdcClusters Clusters (max_values: None, max_size: None, mode: Measured)
+	// Storage: DdcClusters ClustersNodesStats (r:1 w:1)
+	// Proof Skipped: DdcClusters ClustersNodesStats (max_values: None, max_size: None, mode: Measured)
+	// Storage: DdcClusters ClustersNodes (r:1 w:1)
+	// Proof Skipped: DdcClusters ClustersNodes (max_values: None, max_size: None, mode: Measured)
+	fn validate_node() -> Weight {
+		Weight::from_parts(401_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/weights.rs:113]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

File: pallets/ddc-staking/src/weights.rs [hunk 9]
```diff
@@ -35,153 +35,306 @@ pub trait WeightInfo {
...
+	// Proof Skipped: DdcStaking Providers (max_values: None, max_size: None, mode: Measured)
 	// Storage: Balances Locks (r:1 w:1)
+	// Proof: Balances Locks (max_values: None, max_size: Some(1299), added: 3774, mode: MaxEncodedLen)
 	// Storage: System Account (r:1 w:1)
+	// Proof: System Account (max_values: None, max_size: Some(128), added: 2603, mode: MaxEncodedLen)
 	// Storage: DdcStaking LeavingStorages (r:1 w:0)
+	// Proof Skipped: DdcStaking LeavingStorages (max_values: None, max_size: None, mode: Measured)
 	fn withdraw_unbonded() -> Weight {
-		Weight::from_parts(33_000_000_u64, 0)
+		Weight::from_parts(565_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-staking/src/weights.rs:91]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

File: pallets/ddc-staking/src/weights.rs [hunk 10]
```diff
@@ -35,153 +35,306 @@ pub trait WeightInfo {
...
+	// Proof Skipped: DdcStaking Providers (max_values: None, max_size: None, mode: Measured)
 	// Storage: DdcNodes StorageNodes (r:1 w:0)
+	// Proof Skipped: DdcNodes StorageNodes (max_values: None, max_size: None, mode: Measured)
 	// Storage: Balances Locks (r:1 w:1)
+	// Proof: Balances Locks (max_values: None, max_size: Some(1299), added: 3774, mode: MaxEncodedLen)
 	// Storage: System Account (r:1 w:1)
+	// Proof: System Account (max_values: None, max_size: Some(128), added: 2603, mode: MaxEncodedLen)
 	fn unbond() -> Weight {
-		Weight::from_parts(37_000_000_u64, 0)
+		Weight::from_parts(686_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-staking/src/weights.rs:76]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

File: pallets/ddc-staking/src/weights.rs [hunk 11]
```diff
@@ -35,153 +35,306 @@ pub trait WeightInfo {
...
 	}
 	// Storage: DdcStaking Ledger (r:1 w:1)
+	// Proof Skipped: DdcStaking Ledger (max_values: None, max_size: None, mode: Measured)
 	// Storage: DdcStaking Storages (r:1 w:1)
+	// Proof Skipped: DdcStaking Storages (max_values: None, max_size: None, mode: Measured)
 	// Storage: DdcClusters ClustersGovParams (r:1 w:0)
+	// Proof Skipped: DdcClusters ClustersGovParams (max_values: None, max_size: None, mode: Measured)
 	fn chill() -> Weight {
-		Weight::from_parts(28_000_000_u64, 0)
+		Weight::from_parts(394_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-staking/src/weights.rs:119]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

File: pallets/ddc-staking/src/weights.rs [hunk 12]
```diff
@@ -35,153 +35,306 @@ pub trait WeightInfo {
...
+	// Proof Skipped: DdcStaking Nodes (max_values: None, max_size: None, mode: Measured)
 	// Storage: DdcStaking Providers (r:1 w:1)
+	// Proof Skipped: DdcStaking Providers (max_values: None, max_size: None, mode: Measured)
 	// Storage: DdcStaking Storages (r:1 w:0)
+	// Proof Skipped: DdcStaking Storages (max_values: None, max_size: None, mode: Measured)
 	// Storage: DdcStaking LeavingStorages (r:1 w:0)
+	// Proof Skipped: DdcStaking LeavingStorages (max_values: None, max_size: None, mode: Measured)
 	fn set_node() -> Weight {
-		Weight::from_parts(14_000_000_u64, 0)
+		Weight::from_parts(296_000_000_u64, 0)
```
Comment (yahortsaryk) [inline pallets/ddc-staking/src/weights.rs:290]: The weights were taken from Rakan's PR and was executed on a dedicated benchmarking machine

**Review by MRamanenkau — APPROVED** [review]

**Pull request merged**